lib/api/src/backend/sys/async_runtime.rs

```diff
@@ -85,86 +110,107 @@ where
 
-pub(crate) fn call_function_async<'a, S>(
-    function: SysFunction,
-    store: &'a mut S,
-    params: Vec<Value>,
-) -> AsyncCallFuture<'a, S>
-where
-    S: AsStoreMut + 'static,
-{
-    AsyncCallFuture::new(function, store, params)
-}
+                // Leave the coroutine context since we're yielding back to the
+                // parent stack, and will be inactive until the future is ready.
+                ctx_ref.leave();
 
-enum AsyncYield {
-    HostFuture(HostFuture),
+                // Now we can yield back to the runtime while we wait
+                let result = ctx_ref
+                    .block_on_future(Box::pin(future))
+                    .map_err(AsyncRuntimeError::RuntimeError);
+
+                // Once the future is ready, we borrow again and restore the current
+                // coroutine.
+                ctx_ref.enter();
+
+                result
+            }
+        }
+    })
 }
 
-enum AsyncResume {
-    Start,
-    HostFutureReady(Result<Vec<Value>, RuntimeError>),
+thread_local! {
+    static CURRENT_CONTEXT: RefCell<Vec<*const CoroutineContext>> = const { RefCell::new(Vec::new()) };
 }
 
-struct AsyncContextState {
+struct CoroutineContext {
     yielder: *const Yielder<AsyncResume, AsyncYield>,
 }
 
-impl AsyncContextState {
+impl CoroutineContext {
     fn new(yielder: &Yielder<AsyncResume, AsyncYield>) -> Self {
         Self {
             yielder: yielder as *const _,
         }
     }
 
-    fn enter(&self) -> AsyncContextGuard {
+    fn enter(&self) {
         CURRENT_CONTEXT.with(|cell| {
-            let previous = cell.replace(self as *const _);
-            AsyncContextGuard { previous }
+            let mut borrow = cell.borrow_mut();
+
+            // Push this coroutine on top of the active stack.
+            borrow.push(self as *const _);
         })
     }
 
+    // Note: we don't use a drop-style guard here on purpose; if a panic
+    // happens while a coroutine is running, CURRENT_CONTEXT will be in
+    // an inconsistent state. corosensei will unwind all coroutine stacks
+    // anyway, and if we had a guard that would get dropped and try to
+    // leave its context, it'd panic again at the assert_eq! below.
+    fn leave(&self) {
+        CURRENT_CONTEXT.with(|cell| {
+            let mut borrow = cell.borrow_mut();
+
+            // Pop this coroutine from the active stack.
+            assert_eq!(
+                borrow.pop(),
+                Some(self as *const _),
+                "Active coroutine stack corrupted"
+            );
+        });
+    }
+
+    fn get_current() -> Option<*const Self> {
+        CURRENT_CONTEXT.with(|cell| cell.borrow().last().copied())
+    }
+
     fn block_on_future(&self, future: HostFuture) -> Result<Vec<Value>, RuntimeError> {
-        let yielder = unsafe { &*self.yielder };
+        let yielder = unsafe { self.yielder.as_ref().expect("yielder pointer valid") };
```

Comment:
Should this return an `AsyncRuntimeError` yield error?

Reply:
not really, if we have a null yielder, it's an unrecoverable internal error